### ISSUE TITLE
Overwriting of user in candidate portal if candidate stays logged in during changes

### ISSUE THREAD

**Issue by cazcam34** [talentbeyondboundaries/tbbtalentv2#1163]
UPDATE: This actually can't happen (despite appearing to look like it would in the code) as all changes saved to the user object will flush the cache and fetch the updated one in authorisation filter. So the portal is always going to have an updated loggedInUser. However.... I did experience this problem when I was directly updating the partner of a user on the DB, as this doesn't call the save() method and evict the cache. So we need to be careful making direct DB updates, as already established for redis cache flushing in PR #1138.

Similar to issue #1137 where the candidate object was getting overwritten, there is also at least one case of the user being overwritten. Eg. it can reset a partner change back to the old partner

One example below in updatePersonal(), the user that is saved to the database is the outdated loggedInUser and not the updated user object:
<img width="1808" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMS9DQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--68b83ce335a88d4c4d02ce0baf4c1ced726ec1a9/Screen%20Shot%202024-07-09%20at%204.10.04%20pm.png" alt="Screen Shot 2024-07-09 at 4.10.04 pm.png" />

**Comment by cazcam34** [comment]
Another place it is done (method looks very similar to updatePersonal()... need to check if these are both used):
<img width="1228" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMlhDQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--139fad5d23581591047ccdff40b2e45c4238b45b/Screen%20Shot%202024-07-09%20at%204.16.18%20pm.png" alt="Screen Shot 2024-07-09 at 4.16.18 pm.png" />

**Comment by cazcam34** [comment]
Another example if a candidate updates an email address:
<img width="1464" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMmJDQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9b0d35baf579150fb36a1e7f1c709b3976fb48a8/Screen%20Shot%202024-07-09%20at%204.20.35%20pm.png" alt="Screen Shot 2024-07-09 at 4.20.35 pm.png" />

**Comment by cazcam34** [comment]
NOTE! After more testing I can see that this overwrite only seemed to occur on certain fields (the intake fields, the partner field on the user). However things like gender, names, emails those do not get overwritten...
After some digging I found out how/where the logged in user is cached. This is the repository call that we do in:
<img width="1190" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN1hDQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--5533a86955bb0112deda30c67bd67c98ba620b72/Screen%20Shot%202024-07-09%20at%204.58.41%20pm.png" alt="Screen Shot 2024-07-09 at 4.58.41 pm.png" />

Also on save the userRepository gets rid of the cache, so perhaps that is resetting the cache object and would explain why certain fields aren't being cached... by still don't understand why some are working and others arent!

So far the user object can be replaced by the loggedInUser object if the candidate updates:
- First name
- last name
- email

**Comment by cazcam34** [comment]
@sadatmalik @samschlicht Some late night digging led me to find a bit more information. I think this is actually not a bug here.... although it appeared to be one. I'm just writing this as a record of where I am at but thought you might be interested also. 

There are instances where it appears the bug SHOULD appear as per the bug that we identified in issue #1137 but it doesn't. For examples see screenshots above where we are updating a loggedInUser as opposed to a fetched user object. What was confusing was that when looking at the loggedInUser object it actually already updated! And it contained the changes that were made in the admin portal! 

Digging deeper I found that on the save method of the userRepository there is a cache evict. So in this case, the admin portal updates a users email, this saves to the repository and evicts the cache. Looking into this more I see its a [Spring annotation](https://www.baeldung.com/spring-boot-evict-cache) so I do think it could tie in with the Spring authorisation security.
<img width="1074" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeWZFQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--c8a9ce0d4f58cb91b1365ed1977f3f9d85ccbc7f/Screen%20Shot%202024-07-09%20at%208.55.56%20pm.png" alt="Screen Shot 2024-07-09 at 8.55.56 pm.png" />

Then on the candidate portal (already logged in prior to the admin portal change) for every page visit (e.g click on the edit tab for a section) we are calling the loadUserbyUsername() method, and this appears to be coming from doFilterInternal when I step the debug back. I don't understand yet what this doFilterInternal does or why it is called, but can see it is called when editing a candidate profile on the candidate portal.
<img width="1729" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeVhFQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--12383301e1b9479d21b491d61e36274953967a4e/Screen%20Shot%202024-07-09%20at%208.56.11%20pm.png" alt="Screen Shot 2024-07-09 at 8.56.11 pm.png" />

<img width="2000" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeXJFQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0a018d0768f6d6abd70bac0e82d989f7aef4ad9c/Screen%20Shot%202024-07-09%20at%208.56.04%20pm.png" alt="Screen Shot 2024-07-09 at 8.56.04 pm.png" />

Calling this method, after evicting the cache I then am fetching an updated user object to use as my loggedInUser object. So then when I go to update the user, I can see the changes are already on the loggedInUser object!

HOWEVER if for some reason the loggedInUser is different without preceeding it with a userRepository save action (e.g. the database was manually edited) then this loggedInUser won't be up to date. This is what I was doing to change the partner, directly on the DB and why it was reverting.

Summary:
- The loggedInUser/candidate gets updated after a user object is saved, as it is evicted and eventually fetched again.
- If we do something to the candidate only on the admin portal, then the cache and the DB will be out of sync. This is why we have added protocols for the candidate updates to occur on the updated DB object as opposed to the loggedIn candidate object. (issue #1137)
- If we do something to only the candidate object on the admin portal, and then do something to the loggedInUser see examples above, does that affect anything? A: The save returns the updated object user object, so in the code we either get the candidate off that updated object (this being the updated candidate object) or if we are making changes we fetch the candidate object from the DB anyways. So there are safeguards, however technically the cached loggedInUser is still out of date, and will stay like this until the user logs out or the user performs an action that will save the user object (update any user fields).

Is there a potential use for using cacheEvict on the candidate repository save also, as a quicker fix to the out of sync issue?
